### DISCUSSION TITLE
polydata faces and delaunay_2d

### DISCUSSION THREAD

**Discussion by katya-stanzy** [pyvista/pyvista#4638] [Q&A] [answered]
Hello, I would like to create a surface with an enforced (hand-made) structure of faces using arrays of points and faces as below: 

[faces.txt](https://github.com/pyvista/pyvista/files/11994932/faces.txt)
[points.txt](https://github.com/pyvista/pyvista/files/11994933/points.txt)

and a code:
>>>surf = pv.PolyData(var_inp=points, faces=faces)

However, when further creating a triangulated surface as 
>>>surf.delaunay_2d() 

the faces are ignored,  and running 
>>>surf.delaunay_2d(edge_source=surf) 

crashes the kernel.

Can you please help locate the problem?

**Comment by darikg** [comment]
- one of the faces references point index 40 (ie the 41st point), but there's only 40 points
  - delaunay_2d works by projecting the 3d points onto a 2d plane and triangulating there but it's not clear this will result in anything meaningful from your 3d mesh

**Reply by katya-stanzy** [reply]
Hi darikg -- thank you for the reply. 

- Yes, having No 40 in the middle of the list definitely was not helpful! Thank you. 
- However, I am still puzzled by your 2nd comment. My points are arranged as rail tracks: 20 (nn 0-19) on the left and 20 (nn 20-39) on the right. The triangles are created so as to cover the whole surface between the 'tracks'. I'd like to drill a bit deeper into the question because .delaunay_2d is still struggling with the way my data is set out. What needs to change? Should neighboring triangles be listed one after the other? Should the order of data points be different from the current one?

![Screenshot 2023-07-10 130808](https://github.com/pyvista/pyvista/assets/80893754/1cba7a03-559d-43e5-8eff-e219a4825e07)
![Screenshot 2023-07-10 130909](https://github.com/pyvista/pyvista/assets/80893754/e110d4b3-0656-495b-9b75-fcc741987308)
[faces.txt](https://github.com/pyvista/pyvista/files/12001258/faces.txt)
[points.txt](https://github.com/pyvista/pyvista/files/12001257/points.txt)

**Reply by darikg** [reply]
Here's what I see:

```python
import pyvista as pv
from numpy import loadtxt
pv.set_jupyter_backend('trame')
points = loadtxt('points.txt')
faces = loadtxt('faces.txt').astype('int')
mesh = pv.PolyData(points, faces=faces)
mesh.plot(show_edges=True, opacity=0.5)
```

![Capture](https://github.com/pyvista/pyvista/assets/6875882/a1e18703-cd93-418d-929e-ab5efd9abc6a)

Delaunay_2d projects all the points to a plane and does the triangulation there:

```python
mesh.project_points_to_plane(normal=[0, 1, 0]).extract_all_edges().plot(line_width=1)
```

![Capture2](https://github.com/pyvista/pyvista/assets/6875882/55055800-fb65-476f-b118-b9cf875d2545)

There's still some edges crossing in the projection, so there's no valid triangulation. Maybe there's still some errors in your face array? From the structure you describe, triangles between two tracks, the original mesh should look like a ribbon with holes in it, rather than a tube with holes in it. Then it should be possible to find the correct normal vector such that the triangles project to a valid 2d triangulation.

**Reply by katya-stanzy** [reply]
These are very telling plots! Yes, I can see what went wrong with the triangles. Hopefully, this will be the solution. I’m very grateful.

**Reply by katya-stanzy** [reply]
darikg, thanks again for your help -- I've sorted the crossover between the adjacent triangles hoping that would be it. Unfortunately, the pv.PolyData still produces extra faces across several points ([29, 9, 12] and [9, 29, 32]) that are not in the input.  And the resultant delaunay surface still goes all mad. It ignores the first half of the triangles altogether. (note, the triangles are created in the anticlockwise order hoping that it would help calculate the correct direction of face normals). Extra triangulation or stripping PolyData does not change anything.

Is there something about parameters for either PolyData or delaunay_2d that I should be setting?

![screenshot](https://github.com/pyvista/pyvista/assets/80893754/bc36a366-599a-4196-b1eb-2b64d633c11c)
![Screenshot 2023-07-11 110337](https://github.com/pyvista/pyvista/assets/80893754/d1624fc5-814f-4599-8e52-0409a9b86100)
[points.txt](https://github.com/pyvista/pyvista/files/12012899/points.txt)
[faces.txt](https://github.com/pyvista/pyvista/files/12012900/faces.txt)

**Comment by darikg** [comment]
Hi @katya-stanzy, I think I might be misunderstanding something. The points and faces files you attached are from before the attempted delaunay_2d, correct? At what point do the "extra faces across several points ([29, 9, 12] and [9, 29, 32]) that are not in the input" get introduced?

**Reply by katya-stanzy** [reply]
Yes, it is before the delaunay_2d. These two faces are not among the triangles that I programmed but they are once you plot the PolyData. Very strange.

**Reply by darikg** [reply]
Can you post the code that constructs the point & face arrays and the PolyData?

**Reply by katya-stanzy** [reply]
The points come from real data (an MRI tracing) as two lists, right and left, which are concatenated to get an array of 40 points (i.e. the same data as I have already posted). 

I have since tried this for faces:
order = []
i = 0
while i < sz and i+1+20 < 40:
    a = [3, i, i+20, i+1+20]
    order.append(a)
    i += 1
i+=1
while i+1 < 40:
    a = [3, i+1, i+1-20, i-20]
    order.append(a)
    i += 1

This, for a strip:
strip = list([40, 20, 0, 21])
left = list(range(1,20))
right = list(range(22,40))
for i in range(len(right)):
    strip.append(left[i])
    strip.append(right[i])
strip.append(left[-1])

And this for lines:
left = list(range(0,20)) 
right = list(range(20,40))
z = []
for i in range(0,19):
    z.append(left[i])
    z.append(right[i+1])
lines = left + right[::-1] + z
lines.insert(0, len(lines))

**Reply by darikg** [reply, marked as answer]
IIUC you have two sets of points $l_1, l_2, ..., l_n$ and $r_1, r_2, ..., r_n$ and you want to enforce edges $(l_i, l_{i+1}), (r_i, r_{i+1})$ and $(l_i, r_i)$.

One issue is that 8 of the points are duplicates:
```python
np.unique(points, axis=0).shape[0]
32
```
causing the tracks to loop back on themselves:

```python
def plot_track(plotter, points, color, idx_offset=0):
    plotter.add_lines(points, color=color, width=3, connected=True)
    plotter.add_point_labels(points, labels=[str(i + idx_offset) for i in range(len(points))])

pl = pv.Plotter()
plot_track(pl, points[:20], 'red')
plot_track(pl, points[20:], 'blue')
pl.show()
```

![1](https://github.com/pyvista/pyvista/assets/6875882/2a6fef5d-4e01-48aa-8be3-a747a080995d)

A lazy way out of this would be to `clean` the points, discarding non-unique points:
```python
tolerance = 0.1
left = pv.PolyData(points[:20]).clean()
right = pv.PolyData(points[20:]).clean()
pl = pv.Plotter()
plot_track(pl, left.points, 'red')
plot_track(pl, right.points, 'blue')
pl.show()
```

![2](https://github.com/pyvista/pyvista/assets/6875882/5a92c696-cf69-491f-9a72-cc7a5595429c)

Luckily we still have the same number of points in each track. Instead of manually constructing triangles, it's a lot easier to construct quads $(l_i, l_{i+1}, r_{i+1}, r_i)$, and then let `triangulate()` handle splitting them into 2 triangles each:

```python
n = left.n_points
assert right.n_points == n
faces = np.array([[4, i, i + 1, i + n + 1, i + n] for i in range(n - 1)])
mesh = pv.PolyData(np.vstack([left.points, right.points]), faces=faces)
mesh = mesh.triangulate()
mesh.plot(show_edges=True)
```

![3](https://github.com/pyvista/pyvista/assets/6875882/2f62f2e0-9dc8-4916-b74e-26925f64f7df)

No need for delaunay_2d! If there were different numbers of points between tracks, you'd have to drop the edge constraint $(l_i, r_i)$. Construct one single face looping around the entire thing, then do delaunay triangulation:

```python
nl = left.n_points
nr = right.n_points
faces = np.concatenate([[nl + nr], np.arange(nl), nl + np.arange(nr)[::-1]])  # One face with nl + nr vertices
orig = pv.PolyData(np.vstack([left.points, right.points]), faces=faces)
triangulated = orig.delaunay_2d(edge_source=orig)
triangulated.plot(show_edges=True)
```

**Reply by katya-stanzy** [reply]
Hi darikg -- thank you! I did not realise that my original data contained duplicates -- kind of silly, really -- but it is also nice to know that there is a useful .clean() function.

The dataset you created with the help of quadrilaterals serves my purpose perfectly. Also, it does not have any problem with the direction of the normals (I think my manual triangulation did).

Thank you again.
